src/spi.rs

```diff
@@ -0,0 +1,495 @@
+});
+
+impl_pins_trait!(2 => {
+    TRAIT: SCK,
+    AF: AF5,
+    PINS: [PB10, PB13,]
+});
+impl_pins_trait!(2 => {
+    TRAIT: MISO,
+    AF: AF5,
+    PINS: [PB14, PC2,]
+});
+impl_pins_trait!(2 => {
+    TRAIT: MOSI,
+    AF: AF5,
+    PINS: [PB15, PC3,]
+});
+
+impl_pins_trait!(3 => {
+    TRAIT: SCK,
+    AF: AF6,
+    PINS: [PC10,]
+});
+impl_pins_trait!(3 => {
+    TRAIT: MISO,
+    AF: AF6,
+    PINS: [PC11,]
+});
+impl_pins_trait!(3 => {
+    TRAIT: MOSI,
+    AF: AF6,
+    PINS: [PC12,]
+});
+
+#[cfg(feature = "STM32L476VG")]
+impl_pins_trait!(1 => {
+    TRAIT: SCK,
+    AF: AF5,
+    PINS: [PG9,]
+});
+#[cfg(feature = "STM32L476VG")]
+impl_pins_trait!(1 => {
+    TRAIT: MISO,
+    AF: AF5,
+    PINS: [PG10,]
+});
+#[cfg(feature = "STM32L476VG")]
+impl_pins_trait!(1 => {
+    TRAIT: MOSI,
+    AF: AF5,
+    PINS: [PG11,]
+});
+
+//Reference: Ch. 42.4.7 Configuration of SPI
+///Describes raw SPI from device crate
+pub trait InnerSpi where Self: Sized {
+    ///Index of SPI, used at runtime to verify that correct PIN is used.
+    const IDX: u8;
+    ///Type of APB used by SPI.
+    type APB;
+
+    ///Retrieves Clocks frequency corresponding to SPI.
+    fn get_clock_freq(clocks: &Clocks) -> Hertz;
+
+    ///Retrieves CR1 register block.
+    fn cr1(&self) -> & ::stm32l4x6::spi1::CR1;
+
+    ///Retrieves CR2 register block.
+    fn cr2(&self) -> & ::stm32l4x6::spi1::CR2;
+
+    ///Retrieves SR register block.
+    fn sr(&self) -> & ::stm32l4x6::spi1::SR;
+
+    ///Retrieves DR register block.
+    fn dr(&self) -> & ::stm32l4x6::spi1::DR;
+
+    ///Configures CR1 register
+    fn configure_cr1(&self, freq: Hertz, clocks: &Clocks, mode: Mode) {
+        let br = match Self::get_clock_freq(clocks).0 / freq.0 {
+            0 => unreachable!(),
+            1...2 => 0b000,
+            3...5 => 0b001,
+            6...11 => 0b010,
+            12...23 => 0b011,
+            24...39 => 0b100,
+            40...95 => 0b101,
+            96...191 => 0b110,
+            _ => 0b111,
+        };
+
+        self.cr1().write(|w| unsafe {
+            w.br().bits(br)
+             .cpol().bit(mode.polarity == Polarity::IdleHigh)
+             .cpha().bit(mode.phase == Phase::CaptureOnSecondTransition)
+             //2-line undirectional for Master mode
+             .bidimode().clear_bit()
+             .lsbfirst().clear_bit()
+             //TODO: CRC option?
+             .crcen().clear_bit()
+             .ssi().set_bit()
+             .ssm().set_bit()
+             .mstr().set_bit()
+        });
+    }
+
+    ///Configures CR2 register
+    fn configure_cr2(&self) {
+        self.cr2().write(|w| unsafe {
+            //Data size 8 bit
+            w.ds().bits(0b111)
+             .ssoe().set_bit()
+             //RXNE event is generated if the FIFO level is greater than or equal to 1/4 (8-bit)
+             .frxth().set_bit()
+        });
+    }
+
+    ///Enables SPI
+    fn enable(apb: &mut Self::APB);
+}
+
+impl InnerSpi for SPI1 {
```

Comment:
This should be macro-ized to reduce code duplication and handle cases where different chips have different SPI busses (do any variants of the STM32L4x6 family have 1 or 2 SPI busses? 4? etc)

Reply:
Well I did put common part of code in a separate default methods.
I jsut don't feel there is a need to do the macros for the rest?

Also I'm not sure about what do you mean by several busses?
It seems to use only APB2 and APB1 afaik